text/0000-undo-universal-impl-trait.md

```diff
@@ -0,0 +1,309 @@
+    actually pretty simple to wrap your fingers around.
+  - The argument of symmetry (having `impl Trait` in argument position mirroring the `impl Trait` in
+    return position) doesn’t seem to account for history: we’ve been using Rust without `impl Trait`
+    in argument position for years now and no one has ever felt the need to mirror existential
+    quantification with universal quantification in argument position.
+  - People can already write a lot of code without even needing polymorphic code / universal
+    quantification, especially newcomers. Education / teching about type variables and universal
+    quantification can be postponed to the end of the book if the authors are afraid it’s too hard
+    for newcomers.
+  - It seems like proficient developers have no reason to use `impl Trait` in argument position, so
+    why encourage newcomers to do so? What is the **real value** added by such a feature?
+  - About the [dialectical ratchet], it may apply to references / lifetimes but **not** to
+    `impl Trait` for argument position. The concepts of references and lifetimes complement each
+    other, they are used together, not only by newcomers. **These are orthogonal concepts**. Having
+    newcomers first learn `impl Trait` for arguments to express universally quantified generics
+    means that they still will have to unlearn or add a learning exception later when they get
+    introduced to type variables, which might confuse them because of two overlapping distinct
+    constructs to actually express the same thing – plus the former is strictly less powerful. It’s
+    a bit akin to `<T: Trait>` vs. `<T> where T: Trait`. You’re firstly introduced to the former
+    then understand the second one is more powerful (because you cannot express bounds on `Self`
+    with the former). So people actualy use `where` clauses instead of the former, and they’re
+    right. Why would you bother learning two ways to express the same thing when one of them is the
+    the other augmented with a bit more power – bounds on `Self`?
+  - The Rust language is already a very, very complex language with a lot of concepts people **will
+    have to learn**, should it be through the book or any other tutorials. Making Rust easy to
+    **use** and easy to **remember** is as important as making it easy to learn and making so is via
+    the documentation, tutorials and the book, not via the language design itself. We can improve
+    the documentation about generics without affecting the current design.
+  - Overlapping concepts harm the consistency and orthoganality of a language. Having several ways
+    to do something forces people to know all the ways – even if they always prefer using one –
+    because they will maintain and contribute to codebases they haven’t read nor written code for
+    before.
+  - Turbofishing is impossible with `impl Trait`, forcing you to use type ascription in other
+    places.
```

Comment:
This is fixable.
Technically, nothing prevents supporting turbofish and making arg position `impl Trait` purely a sugar for named type parameters.

Reply:
It's slightly more arcane, though. Often when doing something like `fn x<A: TraitA, B: TraitB<A>>(b: B)` I write it with the thing the 'main' generic depends on first (I don't know if you are allowed to put it second, but either way it feels 'wrong' to do it in reverse). This might not be so with impl Trait turbofishing, or at any rate is more difficult to figure out.